src/D2L.CodeStyle.Analyzers/Common/MutabilityInspector.cs

```diff
@@ -75,84 +75,94 @@ public sealed class MutabilityInspector {
 		) {
-			if( type is IErrorTypeSymbol  || type == null ) {
-				throw new Exception( $"Type '{type}' cannot be resolved. Please ensure all dependencies " 
+			if( type is IErrorTypeSymbol || type == null ) {
+				throw new Exception( $"Type '{type}' cannot be resolved. Please ensure all dependencies "
 					+ "are referenced, including transitive dependencies." );
 			}
 
 			if( IsTypeKnownImmutable( type ) ) {
-				return false;
+				return MutabilityInspectionResult.NotMutable();
 			}
 
 			if( type.TypeKind == TypeKind.Array ) {
-				return true;
+				return MutabilityInspectionResult.Mutable( null, type.GetFullTypeName(), MutabilityTarget.Type, MutabilityCause.IsAnArray );
 			}
 
 			if( !flags.HasFlag( MutabilityInspectionFlags.IgnoreImmutabilityAttribute ) && IsTypeMarkedImmutable( type ) ) {
-				return false;
+				return MutabilityInspectionResult.NotMutable();
 			}
 
 			if( typeStack.Contains( type ) ) {
 				// We have a cycle. If we're here, it means that either some read-only member causes the cycle (via IsMemberMutableRecursive), 
 				// or a generic parameter to a type causes the cycle (via IsTypeMutableRecursive). This is safe if the checks above have 
 				// passed and the remaining members are read-only immutable. So we can skip the current check, and allow the type to continue 
 				// to be evaluated.
-				return false;
+				return MutabilityInspectionResult.NotMutable();
 			}
 
 			typeStack.Add( type );
 			try {
 
 				if( ImmutableContainerTypes.Contains( type.GetFullTypeName() ) ) {
 					var namedType = type as INamedTypeSymbol;
-					bool isMutable = namedType.TypeArguments.Any( t => IsTypeMutableRecursive( t, MutabilityInspectionFlags.Default, typeStack ) );
-					return isMutable;
+					foreach( var arg in namedType.TypeArguments ) {
+						var result = InspectTypeRecursive( arg, MutabilityInspectionFlags.Default, typeStack );
+						if( result.IsMutable ) {
+							// modify the result to target the type argument
+							result = MutabilityInspectionResult.Mutable( result.MemberPath, result.TypeName, MutabilityTarget.TypeArgument, result.Cause.Value );
+							return result;
+						}
+					}
+					return MutabilityInspectionResult.NotMutable();
 				}
 
 				if( type.TypeKind == TypeKind.Interface ) {
-					return true;
+					return MutabilityInspectionResult.Mutable( null, type.GetFullTypeName(), MutabilityTarget.Type, MutabilityCause.IsAnInterface );
 				}
 
 				if( !flags.HasFlag( MutabilityInspectionFlags.AllowUnsealed )
 						&& type.TypeKind == TypeKind.Class
 						&& !type.IsSealed
 					) {
-					return true;
+					return MutabilityInspectionResult.Mutable( null, type.GetFullTypeName(), MutabilityTarget.Type, MutabilityCause.IsNotSealed );
 				}
 
-				foreach( ISymbol member in type.GetNonStaticMembers() ) {
-					if( IsMemberMutableRecursive( member, MutabilityInspectionFlags.Default, typeStack ) ) {
-						return true;
+				foreach( ISymbol member in type.GetExplicitNonStaticMembers() ) {

```

Comment:
This change was required because when I started asserting causes in the tests, one of the tests was raising a diagnostic on a different cause than was expected (it was looking at backing fields for properties, where only the property really matters).